### ISSUE TITLE
ResultAsync is not a valid async function return type

### ISSUE THREAD

**Issue by HyperCharlie** [supermacro/neverthrow#222]
After a lot of research into error handling in TypeScript, I've pretty much settled on this library as being the best option for our use. However, I am running into a major blocker. My existing codebase is built using async functions and await, but after converting the return type of a method from a Promise<something> to ResultAsync<something, someError>, the compiler complains that 

`Type 'typeof ResultAsync' is not a valid async function return type in ES5/ES3 because it does not refer to a Promise-compatible constructor value.`

I am targeting ES5 in my tsconfig, because this is going to run in a browser, but it doesn't make sense to me that it would not be promise compatible. This seems very basic so I figure I must be missing something here; is there a config option or something I may be overlooking? For the record, I'm using neverthrow 3.1.2. 

As an example, even this basic method throws an error:

`public async testAsync(): ResultAsync<string, Error> {
    return okAsync("Hello world!");
  }`

**Comment by paduc** [comment]
Hi @HyperCharlie !

When you return a `ResultAsync`, your method becomes sync so you should not prefix with `async`.

**Comment by HyperCharlie** [comment]
Thanks for the quick response! I was just coming back to post that I figured it out, that it was a pattern problem. My current code still has a lot of await somePromise() in it that I am mixing with other functions that return ResultAsync, and all I need to do is wrap the legacy functions with .fromPromise to make it work. 

Fortunately .map() is chainable; many of my methods need 3 or 4 async results that I use Promise.all() to bunch together*. I would use neverthrow.combine() to emulate this, but according to the docs it does not support heterogenous lists, which I would need to avoid chaining .map(). Any word on when support for that feature is expected?

* Things like getContext(), getBrowserNode() and getChannelAddress(), which are all needed by the method and are async, but don't depend on one another, so I can use array destructuring and Promise.all() to get them properly asynchronously and cleanly.

**Comment by paduc** [comment]
If you have several async results coming from promises, you can chose to call `ResultAsync.fromPromise` on the `Promise.all` promise to avoid `combine`.

```ts
ResultAsync.fromPromise(Promise.all([getContext(), getBrowserNode(), getChannelAddress()], (e) => new Error('oops'))
           .map(([context, browserNode, channelAddress]) => ... );
```

**Comment by HyperCharlie** [comment]
@paduc good idea, that will help clean things up during the transition!

**Comment by HyperCharlie** [comment]
OK, last question related to patterns- I have now converted most of my methods to returning ResultAsync, but now I need to combine several heterogeneous ResultAsyncs, where a simple .map chain doesn't work. Right now, I see two ways to do it, something like:

```function func1(): ResultAsync<string, Error> {
  return okAsync("Hello!");
}

function func2(): ResultAsync<number, Error> {
  return okAsync(1);
}

function combinedFunc(): ResultAsync<bool, Error> {
  return func1()
    .andThen((myStr) => {
      return func2()
        .map((myNum) => {
          return myNum == myStr;
        }
    }
}

function async combinedFunc2(): ResultAsync<boolean, Error> {
  const func1Res = func1();
  const func2Res = func2();

  return asyncResult.fromPromise(Promise.all([func1Res, func2Res])).map((myVals) => {
    const [myStrRes, myNumRes] = myVals;

    return myStrRes.value == myNumRes.value;
  });
}
  
```

This offends my sense of aesthetics, but I'm not sure of a better pattern. combinedFunc1 seems unnecessarily ugly, combinedFunc2 has these extra jumps in there that are very hard to explain but seems more elegant, is there a third way I may be overlooking?

**Comment by paduc** [comment]
If both `func1()` and `func2()` return `ResultAsync` types that you need to compare, I would create a `ResultAsync` that has them both and then operate the comparison.

```typescript
const res = func1().andThen((value1) => func2().map((value2) =>({ value1, value2 }))) // This combines the two values in a single composite value
                   .map(({ value1, value2}) => value1 == value2) // This does the comparison
```

**Comment by supermacro** [comment]
Combining heterogeneous Results or ResultAsyncs has been on my todo-list for a while. In lieu of this missing functionality in neverthrow, this is how you could do it:

The [`combine`](https://github.com/supermacro/neverthrow#combine) function already does work for heterogeneous lists if you ran this code in javascript. The limitation right now is one of a simplified type definition - which forces you to only work on homogeneous lists.  

Proof: https://repl.it/@gDelgado14/heterogeneous-combine

So what you can do for now is provide typecasts on `combine` function invocations (note that for this to work you need to implement function overloading):

```typescript
import { combine, Result } from 'neverthrow'

combine([ myOneResult, myOtherResult ]) as Result<[ Blah, Whatever ], Err1 | Err2>
```

If you want to get really fancy (which is what I have to implement anyways when I do get around to doing this) is to provide function overloads for all the various invocations of `combine` (just like how [`Promise.all` is implemented](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.es2015.promise.d.ts#L89) in typescript)

**Comment by HyperCharlie** [comment]
Thanks both @paduc and @supermacro. @paduc, your formulation is slick but gets really hard to follow for 3+ results; I just tried rewriting with that and it was not to my liking. Requiring a typecast seems a small price to pay for now @supermacro, I just rewrote the function again that way and it is quite readable.

It's funny you mention Promise.all's TS implementation, that was the first thing that came to mind when I saw that combine() didn't support multiple types yet. I kind of figured that was the reason, what a mess!

**Comment by HyperCharlie** [comment]
Just for anybody who may stumble on this, in order to make the casting version by @supermacro work, I needed to add some extra casts:

''' 
combine([ myOneResult as ResultAsync<any, any>, 
myOtherResult ]) as unknown as ResultAsync<[ Blah, Whatever ], Err1 | Err2>
'''

It was complaining that myOneResult and myOtherResult did not have enough in common; casting one of them to a ResultAsync<any, any> was enough to allow all of them to work. Also, I needed to cast the result through an unknown, for basically the same reasons.

@supermacro, your example had Result as the return value of combine; if you combine ResultAsync combine returns a ResultAsync, correct? Just for completion's sake, if I combine() a Result and a ResultAsync, is that A. legal and B. does it return a ResultAsync?

**Comment by paduc** [comment]
I believe a union type can ease the use of `combine` with heterogeneous types.

Example:
```ts

const a = okAsync<string, Error>('A');
const b = okAsync<number, Error>(1);

// Use the union of all possible types in the combine template
const combination = combine<string | number, Error>([a, b]).map(
  // cast the types of the resultList with the value types of a and b
  ([aValue, bValue]: [string, number]) => {
    // aValue is string
    // bValue is number
  }
);
```

**Comment by HyperCharlie** [comment]
@paduc I gave that a shot, but I was not able to make it work; best I was able to manage gave this error:

```Argument of type '([signer, channelAddress, browserNode]: [ethers.providers.JsonRpcSigner, string, BrowserNode]) => Promise<void>' is not assignable to parameter of type '(t: (string | BrowserNode | JsonRpcSigner)[]) => void | Promise<void>'.
  Types of parameters '__0' and 't' are incompatible.
    Type '(string | BrowserNode | JsonRpcSigner)[]' is not assignable to type '[JsonRpcSigner, string, BrowserNode]'.
      Target requires 3 element(s) but source may have fewer.
```

For reference, here's the top of the method I've been experimenting on as formulated as you suggest:

```
public depositFunds(assetAddress: EthereumAddress, amount: BigNumber): ResultAsync<void, BlockchainUnavailableError> {
    const prerequisites = combine<ethers.providers.JsonRpcSigner | string | BrowserNode, 
    RouterChannelUnknownError | CoreUninitializedError | NodeError | Error | BlockchainUnavailableError>([this.blockchainProvider.getSigner() as ResultAsync<any, any>,
    this.vectorUtils.getRouterChannelAddress(),
    this.browserNodeProvider.getBrowserNode()]);

    return prerequisites
    .map(async ([signer, channelAddress, browserNode]: [ethers.providers.JsonRpcSigner, string, BrowserNode]) => { ...
```

On a tangentially related note, if my .map() callback has an error in it, do I just throw an exception? If I do `return errAsync(new Error())` in the body of the function, the typing blows up completely, which makes sense. I mean, my return value is all of a sudden MyType | Err<unknown, Error>. But if I throw an exception, the type information would be lost. I just read the whole wiki again looking for an example of this but didn't spot it. There are really two separate things going on. First, calling library methods that could throw an error, which I should be able to wrap with fromPromise or fromThrowable. Second though, would be intentional errors that I want to report. In that case, .map would actually have to have a more complex return type, which I don't think it can handle. 

That being said, is the issue that I should use .andThen for any code that may generate errors? I have been struggling with the code I posted above once I convert it to .andThen:
```
return prerequisites.andThen(async (myVals) => {
  ...
  if (badStuff) {
    return errAsync();
  }
  return okAsync(null);
}
```

Results in the following error:

```
Argument of type '(myVals: [JsonRpcSigner, string, BrowserNode]) => Promise<Result<null, CoreUninitializedError | Error | BlockchainUnavailableError | NodeError | RouterChannelUnknownError>>' is not assignable to parameter of type '(t: [JsonRpcSigner, string, BrowserNode]) => Ok<void, CoreUninitializedError | Error | BlockchainUnavailableError | NodeError | RouterChannelUnknownError> | Err<...> | ResultAsync<...>'.
```

**Comment by HyperCharlie** [comment]
I just thought I'd toss this out there- here's my refactored function in all it's glory. It still needs some smarter and/or more fine-grained error handling. I'm trying to decide if there is a better way than massive union types to deal with functions that can return multiple kinds of errors. I mean, I could put a .mapErr() on the end and just return a single error type, like a DepositFundsError, which could event have something like a sourceError member, but that just seems like it obfuscates things. All of the errors in my definition are known, expected types of errors, even if I can't recover from them.

One of the reasons I'm doing this work at all is to better surface the kinds of errors that can occur in our system; the plan is that on a second sweep, we can analyze and put in more and/or better recovery code. This is all blockchain and financial stuff so we are trying to make it as error tolerant as possible.

```
public depositFunds(assetAddress: EthereumAddress, amount: BigNumber): ResultAsync<null, RouterChannelUnknownError | CoreUninitializedError | NodeError | Error | BlockchainUnavailableError> {
    const prerequisites = combine([this.blockchainProvider.getSigner() as ResultAsync<any, any>,
    this.vectorUtils.getRouterChannelAddress(),
    this.browserNodeProvider.getBrowserNode()]) as unknown as ResultAsync<[ethers.providers.JsonRpcSigner, string, BrowserNode], RouterChannelUnknownError | CoreUninitializedError | NodeError | Error | BlockchainUnavailableError>;

    let signer: ethers.providers.JsonRpcSigner | null = null;
    let channelAddress: string | null = null;
    let browserNode: BrowserNode | null = null;

    return prerequisites.andThen((vals) => {
      [signer, channelAddress, browserNode] = vals;

      let tx: ResultAsync<TransactionResponse, BlockchainUnavailableError>;

      if (assetAddress == "0x0000000000000000000000000000000000000000") {
        console.log("Transferring ETH.");
        // send eth
        tx = ResultAsync.fromPromise(signer.sendTransaction({ to: channelAddress, value: amount }),
        (err) => {return err as BlockchainUnavailableError});
      } else {
        console.log("Transferring an ERC20 asset.");
        // send an actual erc20 token
        let tokenContract = new Contract(assetAddress, ERC20Abi, signer);
        tx = ResultAsync.fromPromise(tokenContract.transfer(channelAddress, amount),
        (err) => {return err as BlockchainUnavailableError});
      }

      return tx
    })
    .andThen((tx) => {
       // TODO: Wait on this, break it up, this could take a while
       return ResultAsync.fromPromise(tx.wait());
    })
    .andThen((receipt) => {
      if (browserNode == null || channelAddress == null) {
        return errAsync(new Error("Really screwed up!"));
      }

      return ResultAsync.fromPromise(browserNode.reconcileDeposit({
        assetId: assetAddress,
        channelAddress
      }), (err) => {return err as BlockchainUnavailableError});
    })
    .andThen((depositVal) => {
      // I can not for the life of me figure out why depositVal is coming back
      // as "unknown"
      const depositRes = depositVal as Result<NodeResponses.Deposit, NodeError>;
      if (depositRes.isError) {
        console.log(depositRes.getError());
        return errAsync(depositRes.getError() as NodeError);
      }

      const deposit = depositRes.getValue();

      if (deposit.channelAddress !== channelAddress) {
        return errAsync(new Error("Something has gone horribly wrong!"));
      }

      return okAsync(null);
    });
```

Note that the library that provides BrowserNode (Connext's Vector) returns results objects that are almost, but not quite, NeverThrow's, and in the final .andThen you can see that. Their typing is done dynamically and for the life of me I can't figure out why I need that cast in there; depositVal keeps coming through as "unknown".

For further reference, here was the original function:

```
public async depositFunds(assetAddress: EthereumAddress, amount: BigNumber): Promise<void> {
    const signerPromise = await this.blockchainProvider.getSigner();
    const channelAddressPromise = await this.vectorUtils.getRouterChannelAddress();
    const browserNodePromise = await this.browserNodeProvider.getBrowserNode();

    const [signer, channelAddress, browserNode] = await Promise.all([signerPromise, channelAddressPromise, browserNodePromise]);

    let tx;

    if (assetAddress == "0x0000000000000000000000000000000000000000") {
      console.log("Transferring ETH.");
      // send eth
      tx = await signer.sendTransaction({ to: channelAddress, value: amount });
    } else {
      console.log("Transferring an ERC20 asset.");
      // send an actual erc20 token
      let tokenContract = new Contract(assetAddress, ERC20Abi, signer);
      tx = await tokenContract.transfer(channelAddress, amount);
    }

    // TODO: Wait on this, break it up, this could take a while
    await tx.wait();

    const depositRes = await browserNode.reconcileDeposit({
      assetId: assetAddress,
      channelAddress,
      // publicIdentifier: browserNode.publicIdentifier
    });

    if (depositRes.isError) {
      console.log(depositRes.getError());
      throw new Error("Deposit can not be reconciled");
    }

    const deposit = depositRes.getValue();

    if (deposit.channelAddress !== channelAddress) {
      throw new Error("Something has gone horribly wrong!");
    }
  }
```
